Minimize unused memory allocated in DataChannel.readLoop.

bshimc opened the pull request:
See https://github.com/pion/webrtc/issues/1516

This patch preserves the semantics of the OnMessage handler and is more safe
but less efficient than the patch first described in #1516.

As expected, when run with GOMAXPROCS=1 there is opportunity for re-use and
syn.Pool is pure overhead.  As allocs/op increases due to sync.Pool overhead
but B/op and ns/op improve drastically.

    $ git co datachannel.go && \
      got test -bench=DataChannelSend -run=XXX \
      -benchmem -cpu=1,2,4,8 -count=10 > original.txt
    $ git co datachannel.go && git apply pool.patch && \
      go test -bench=DataChannelSend -run=XXX \
    	-benchmem -cpu=1,2,4,8 -count=10 > patched.txt

    $ benchstat original.txt patched.txt
    name               old time/op    new time/op    delta
    DataChannelSend      1.45µs ±35%    1.24µs ±40%      ~     (p=0.211 n=9+10)
    DataChannelSend-2     885ns ±14%    2221ns ±44%  +150.87%  (p=0.000 n=10+9)
    DataChannelSend-4    5.19µs ±15%    3.27µs ±16%   -37.13%  (p=0.000 n=9+10)
    DataChannelSend-8    8.07µs ±22%    3.97µs ± 4%   -50.78%  (p=0.000 n=10+8)

    name               old alloc/op   new alloc/op   delta
    DataChannelSend      1.67kB ±72%    0.31kB ±12%   -81.55%  (p=0.000 n=9+10)
    DataChannelSend-2    1.80kB ±21%    0.34kB ±14%   -81.14%  (p=0.000 n=10+8)
    DataChannelSend-4    1.47kB ±14%    1.40kB ± 0%    -4.39%  (p=0.031 n=10+8)
    DataChannelSend-8    2.18kB ±14%    1.45kB ± 0%   -33.57%  (p=0.000 n=9+8)

    name               old allocs/op  new allocs/op  delta
    DataChannelSend        6.00 ± 0%      6.80 ±18%   +13.33%  (p=0.008 n=8+10)
    DataChannelSend-2      6.00 ± 0%      8.11 ±36%   +35.19%  (p=0.000 n=10+9)
    DataChannelSend-4      7.00 ± 0%     37.00 ± 0%  +428.57%  (p=0.000 n=10+10)
    DataChannelSend-8      7.70 ± 9%     39.00 ± 0%  +406.49%  (p=0.000 n=10+7)

#### Description

#### Reference issue
Fixes #...

bshimc commented:
@Sean-Der I've addressed all the CI issues I could but I'm not sure how the system is calculating the reduced coverage.  Perhaps it's that Benchmark tests aren't run.

Sean-Der commented:
@bshimc We unfortunately don't have any benchmark infrastructure. It would be really awesome if we did though! Do you have anything to automate catching memory/performance regressions?  

I am fine with with merging this as is, and we can come back to it. If you are using Pion I would love to add you to the organization and get your help creating a better 'perf culture'. That and fuzz testing are our biggest misses right now.

jsmouret commented:
The original patch in #1516 was a lot nicer without the extra copy.
As v3 is on the way and already has plenty of API changes like the Track interface,
could we add the requirement that OnMessage buffer be only used inside the callback?

Sean-Der commented:
Lets do it! @jsmouret  would you mind adding an entry to the 'v3.0.0' release notes for it?

Would you mind doing the original patch @bshimc? Sorry for the churn, I tend to be more cautious but I love seeing community excitement :) 

I still would love to see if we could get some CI and start paying down performance debt.

bshimc commented:
@Sean-Der @jsmouret I appreciate the enthusiasm but I'd recommend holding off on making the aggressive change.  On further reflection the original patch supplied in the issue is poor API design that is begging to be accidentally misused.  The fundamental problem is that we're passing byte buffers around.  Performance sensitive APIs generally pass io.ReadCloser objects around to avoid allocations so I tried taking a page out of the http.Request.Body API where it's the user's responsibility to call .Close().  Most importantly, this allows the user to decide what the lifecycle of the allocated memory should be.  See https://github.com/pion/webrtc/compare/master...bshimc:readcloser for a working example.  Benchmarking that yielded results that suggest that my benchmark is flawed; there are basically no discernable differences between what's shipping now and the rlBufPool implementation in that branch.  I'll try to run a distributed load test to see if there are improvements because I no longer trust my synthetic Benchmark test. This is all along winded way of advising caution.  There are some gaps in my understanding of what's going on that I need to fill.

```diff
diff --git a/datachannelmessage.go b/datachannelmessage.go
index 1e3c63b..7cbafdf 100644
--- a/datachannelmessage.go
+++ b/datachannelmessage.go
@@ -1,10 +1,15 @@
 package webrtc

+import "io"
+
 // DataChannelMessage represents a message received from the
 // data channel. IsString will be set to true if the incoming
 // message is of the string type. Otherwise the message is of
 // a binary type.
 type DataChannelMessage struct {
        IsString bool
-       Data     []byte
+
+       // It is the caller's responsibility to close Data to release memory
+       // resources.
+       Data io.ReadCloser
 }
diff --git a/datachannel.go b/datachannel.go
index ba1269d..d44509d 100644
--- a/datachannel.go
+++ b/datachannel.go
@@ -3,6 +3,7 @@
 package webrtc

 import (
+       "bytes"
        "errors"
        "fmt"
        "io"
@@ -44,7 +45,7 @@ type DataChannel struct {
        // "blob". This attribute controls how binary data is exposed to scripts.
        // binaryType                 string

-       onMessageHandler    func(DataChannelMessage)
+       onMessageHandler    func(*DataChannelMessage)
        openHandlerOnce     sync.Once
        onOpenHandler       func()
        onCloseHandler      func()
@@ -260,13 +261,13 @@ func (d *DataChannel) onClose() {
 // in size. Check out the detach API if you want to use larger
 // message sizes. Note that browser support for larger messages
 // is also limited.
-func (d *DataChannel) OnMessage(f func(msg DataChannelMessage)) {
+func (d *DataChannel) OnMessage(f func(msg *DataChannelMessage)) {
        d.mu.Lock()
        defer d.mu.Unlock()
        d.onMessageHandler = f
 }

-func (d *DataChannel) onMessage(msg DataChannelMessage) {
+func (d *DataChannel) onMessage(msg *DataChannelMessage) {
        d.mu.RLock()
        handler := d.onMessageHandler
        d.mu.RUnlock()
@@ -311,11 +312,33 @@ func (d *DataChannel) onError(err error) {
        }
 }

+// See https://github.com/pion/webrtc/issues/1516
+// nolint:gochecknoglobals
+var rlBufPool sync.Pool
+
+type rlBuf struct {
+       *bytes.Reader
+       b [dataChannelBufferSize]byte
+}
+
+func (r *rlBuf) Close() error {
+       r.Reader = nil
+       rlBufPool.Put(r)
+       return nil
+}
+
 func (d *DataChannel) readLoop() {
        for {
-               buffer := make([]byte, dataChannelBufferSize)
-               n, isString, err := d.dataChannel.ReadDataChannel(buffer)
+               var b *rlBuf
+               if v := rlBufPool.Get(); v != nil {
+                       b = v.(*rlBuf)
+               } else {
+                       b = &rlBuf{}
+               }
+
+               n, isString, err := d.dataChannel.ReadDataChannel(b.b[:])
                if err != nil {
+                       rlBufPool.Put(b)
                        d.setReadyState(DataChannelStateClosed)
                        if err != io.EOF {
                                d.onError(err)
@@ -324,7 +347,8 @@ func (d *DataChannel) readLoop() {
                        return
                }

-               d.onMessage(DataChannelMessage{Data: buffer[:n], IsString: isString})
+               b.Reader = bytes.NewReader(b.b[:n])
+               d.onMessage(&DataChannelMessage{Data: b, IsString: isString})
        }
 }

```


> If you are using Pion I would love to add you to the organization

Sure; I'm not able to make any concrete time or resource commitments (and I'm certainly not a domain expert in this area) but I'm happy to help where possible.  Feel free to add my alter ego (@bshi) to the organization.

bshimc commented:
As expected, the benchmark wasn't representative of what I was testing in issue #1516.  I've corrected that and the results for the fully compatible changes are quite good.  Notably, the original code hits a thresshold where performance falls off a cliff and this change smoothes out said cliff.
```
$ benchstat original.txt option1.txt
name                 old time/op    new time/op    delta
DSend2-8     20.3µs ±51%     3.7µs ± 6%   -81.74%  (p=0.000 n=10+10)
DSend4-8     23.5µs ±34%     3.6µs ± 8%   -84.80%  (p=0.000 n=10+8)
DSend8-8     18.9µs ±35%     5.8µs ±68%   -69.45%  (p=0.000 n=9+10)
DSend16-8    16.8µs ±30%    10.0µs ±24%   -40.77%  (p=0.000 n=10+10)
DSend32-8    710ms ±100%       0ms ±81%  -100.00%  (p=0.035 n=10+9)

name                 old alloc/op   new alloc/op   delta
DSend2-8     15.3kB ±89%     1.4kB ± 0%   -90.59%  (p=0.000 n=9+10)
DSend4-8     41.7kB ±63%     1.4kB ± 1%   -96.58%  (p=0.000 n=10+10)
DSend8-8     45.0kB ±33%     1.4kB ± 2%   -96.83%  (p=0.000 n=9+10)
DSend16-8    34.0kB ±69%     1.4kB ± 1%   -95.77%  (p=0.000 n=10+10)
DSend32-8   37.4MB ±388%     0.0MB ± 4%  -100.00%  (p=0.000 n=10+7)

name                 old allocs/op  new allocs/op  delta
DSend2-8       15.8 ±46%      38.6 ± 2%  +144.30%  (p=0.000 n=10+10)
DSend4-8       27.1 ±48%      38.0 ± 0%   +40.22%  (p=0.000 n=10+9)
DSend8-8       29.3 ±16%      38.0 ± 0%   +29.55%  (p=0.000 n=9+8)
DSend16-8      23.6 ±41%      37.0 ± 0%   +56.78%  (p=0.000 n=10+9)
DSend32-8    19.3k ±100%      0.0k ± 0%      ~     (p=0.178 n=10+7)
```

I recommend merging this in as is and then we can continue discussing further improvements in the issue.  In this latest PR update, I've also added `-bench=.` to the CI `go test` invocation to get coverage for the benchmark tests.  This just executes one run of the benchmark and shouldn't be considered useful for any kind of performance regression uses.

Sean-Der commented:
Merged with 159ba5aca3b96b85f7a653329102ea12046a1b1f

Thank you so much for doing this. Hopefully you can stay involved with Pion, we really need someone who can push us to better in this direction :)

Sean-Der commented:
@bshimc 

The perf improvement here is really exciting! Check out `examples/data-channels-flow-controls` 

**before**
```
2020/11/14 14:46:31 Throughput: 144.323 Mbps
2020/11/14 14:46:32 Throughput: 147.815 Mbps
2020/11/14 14:46:33 Throughput: 147.522 Mbps
2020/11/14 14:46:34 Throughput: 147.551 Mbps
2020/11/14 14:46:35 Throughput: 147.756 Mbps
2020/11/14 14:46:36 Throughput: 145.140 Mbps
2020/11/14 14:46:37 Throughput: 144.695 Mbps
2020/11/14 14:46:38 Throughput: 144.200 Mbps
2020/11/14 14:46:39 Throughput: 144.131 Mbps
2020/11/14 14:46:40 Throughput: 144.841 Mbps
2020/11/14 14:46:41 Throughput: 145.251 Mbps
2020/11/14 14:46:42 Throughput: 145.237 Mbps
```

**after**
```
2020/11/14 14:44:53 Throughput: 204.586 Mbps
2020/11/14 14:44:54 Throughput: 206.091 Mbps
2020/11/14 14:44:55 Throughput: 205.737 Mbps
2020/11/14 14:44:56 Throughput: 205.667 Mbps
2020/11/14 14:44:57 Throughput: 206.387 Mbps
2020/11/14 14:44:58 Throughput: 207.378 Mbps
2020/11/14 14:44:59 Throughput: 208.151 Mbps
2020/11/14 14:45:00 Throughput: 208.919 Mbps
2020/11/14 14:45:01 Throughput: 209.053 Mbps
2020/11/14 14:45:02 Throughput: 209.243 Mbps
2020/11/14 14:45:03 Throughput: 209.546 Mbps
2020/11/14 14:45:04 Throughput: 210.056 Mbps
2020/11/14 14:45:05 Throughput: 210.680 Mbps
2020/11/14 14:45:06 Throughput: 210.781 Mbps
2020/11/14 14:45:07 Throughput: 211.216 Mbps
```

Sean-Der commented:
Playing with that example it will sometimes bomb on throughput for me. It happens without your patch as well though :(
